cpp/include/rapidsmpf/coll/allreduce.hpp

```diff
@@ -0,0 +1,407 @@
+
+    AllReduce(AllReduce const&) = delete;
+    AllReduce& operator=(AllReduce const&) = delete;
+    AllReduce(AllReduce&&) = delete;
+    AllReduce& operator=(AllReduce&&) = delete;
+
+    /**
+     * @brief Destructor.
+     *
+     * @note This operation is logically collective. If an `AllReduce` is locally
+     *       destructed before `wait_and_extract` is called, there is no guarantee
+     *       that in-flight communication will be completed.
+     */
+    ~AllReduce();
+
+    /**
+     * @brief Insert packed data into the allreduce operation.
+     *
+     * @param packed_data The data to contribute to the allreduce.
+     *
+     * @throws std::runtime_error If insert has already been called on this instance.
+     */
+    void insert(PackedData&& packed_data);
+
+    /**
+     * @brief Check if the allreduce operation has completed.
+     *
+     * @return True if all data and finish messages from all ranks have been
+     *         received and locally reduced.
+     */
+    [[nodiscard]] bool finished() const noexcept;
+
+    /**
+     * @brief Wait for completion and extract the reduced data.
+     *
+     * Blocks until the allreduce operation completes and returns the
+     * globally reduced result.
+     *
+     * @param timeout Optional maximum duration to wait. Negative values mean
+     *        no timeout.
+     *
+     * @return The reduced packed data.
+     * @throws std::runtime_error If the timeout is reached.
+     */
+    [[nodiscard]] PackedData wait_and_extract(
+        std::chrono::milliseconds timeout = std::chrono::milliseconds{-1}
+    );
+
+    /**
+     * @brief Check if reduced results are ready for extraction.
+     *
+     * This returns true once the underlying allgather has completed and, if
+     * `wait_and_extract` has not yet been called, indicates that calling it
+     * would not block.
+     *
+     * @return True if the allreduce operation has completed and results are ready for
+     *         extraction, false otherwise.
+     */
+    [[nodiscard]] bool is_ready() const noexcept;
+
+  private:
+    /// @brief Perform the reduction across all ranks for the gathered contributions.
+    [[nodiscard]] PackedData reduce_all(std::vector<PackedData>&& gathered);
+
+    ReduceOperator reduce_operator_;  ///< Reduction operator
+    BufferResource* br_;  ///< Buffer resource for memory normalization
+
+    Rank nranks_;  ///< Number of ranks in the communicator
+    AllGather gatherer_;  ///< Underlying allgather primitive
+
+    bool inserted_{false};  ///< Whether insert has been called
+};
+
+namespace detail {
+
+/**
+ * @brief Host-side elementwise operator working on raw bytes.
+ *
+ * The callable receives pointers to a single element of `accum` and `incoming`
+ * respectively and must combine them in-place into `accum`. The caller
+ * guarantees that both pointers reference `element_size` bytes.
+ */
+using HostByteOp = std::function<void(void* accum_elem, void const* incoming_elem)>;
```

Comment:
Based on the new changes proposed by @wence- [here](https://github.com/rapidsai/rapidsmpf/pull/683/changes#r2610414273), I think its best to change the reduction function to accept `PackedData&`  rather than individual elements here. I think an element-wise `std::function` would be a bit too inefficient. So, I think let's go ahead with Lawrence's suggestion.